internal/envconfig/envconfig.go

```diff
@@ -12,168 +12,499 @@
+//
+//     export MYVAR="a:b,c:d" // map[string]string{"a":"b", "c":"d"}
+//
+// If you need to modify environment variable values before processing, you can
+// specify a custom mutator:
+//
+//     type Config struct {
+//         Password `env:"PASSWORD_SECRET"`
+//     }
+//
+//     func resolveSecretFunc(key, value string) (string, error) {
+//         if strings.HasSuffix(key, "_SECRET") {
+//             return secretmanager.Resolve(value)
+//         }
+//     }
+//
+//     var config Config
+//     ProcessWith(&config, new(OsLookuper), resolveSecretFunc)
+//
+// It is conceptually similar to kelseyhightower/envconfig, with the following
+// major behavioral changes:
+//
+// - Support for specifying a custom lookup function such as a map, which is
+// useful for testing.
+//
+// - Only populates fields if they contain the zero (or nil) value. This means
+// you can pre-initialize a struct and any pre-populated fields will not be
+// overwritten during processing.
+//
+// - Support for interpolation. The default value for a field can be the value
+// of another field.
+//
+// - Support for arbitrary mutators that change/resolve data before type
+// conversion.
 //
-// This can be used with any secret manager that implements the
-// 'secrets.SecretManager' interface.
 package envconfig
 
 import (
 	"context"
-	"crypto/sha1"
+	"encoding"
+	"errors"
 	"fmt"
-	"io/ioutil"
 	"os"
-	"path/filepath"
+	"reflect"
+	"strconv"
 	"strings"
-
-	"github.com/google/exposure-notifications-server/internal/logging"
-	"github.com/google/exposure-notifications-server/internal/secrets"
-	kenvconfig "github.com/kelseyhightower/envconfig"
+	"time"
 )
 
-const (
-	// SecretPrefix is the prefix, that if the value of an env var starts with
-	// will be resolved through the configured secret store.
-	SecretPrefix = "secret://"
+const envTag = "env"
 
-	// FileSuffix is the suffix to use, if this secret path should be written to a file.
-	// only interpreted on environment variable values that start w/ secret://
-	FileSuffix = "?target=file"
+var (
+	ErrInvalidMapItem     = fmt.Errorf("invalid map item")
+	ErrLookuperNil        = fmt.Errorf("lookuper cannot be nil")
+	ErrMissingKey         = fmt.Errorf("missing key")
+	ErrMissingRequired    = fmt.Errorf("missing required value")
+	ErrNotPtr             = fmt.Errorf("input must be a pointer")
+	ErrNotStruct          = fmt.Errorf("input must be a struct")
+	ErrPrivateField       = fmt.Errorf("cannot parse private fields")
+	ErrRequiredAndDefault = fmt.Errorf("field cannot be required and have a default value")
+	ErrUnknownOption      = fmt.Errorf("unknown option")
 )
 
-// BaseConfig is the default base configuration.
-type BaseConfig struct {
-	// SecretsDir is the base directory where secrets are stored.
-	SecretsDir string `envconfig:"SECRETS_DIR" default:"/var/run/secrets"`
+// Lookuper is an interface that provides a lookup for a string-based key.
+type Lookuper interface {
```

Comment:
naming nit (the worst kind of nit): maybe LookerUpper? Or ConfigResolver? Or just ConfigMap, since most programmers associate that with simple key lookup interfaces.

Reply:
Happy to change it, but this is an internal package (only we can depend on it).

Reply:
External package now.